generator/generator.py

```diff
@@ -0,0 +1,277 @@
+import argparse
+import subprocess
+import sys
+import json
+import hashlib
+import shutil
+from pathlib import Path
+
+DOWNLOAD_URL = "https://downloads.arduino.cc/arduino-fwuploader"
+FQBNS = {
+    "mkr1000": "arduino:samd:mkr1000",
+    "mkrwifi1010": "arduino:samd:mkrwifi1010",
+    "nano_33_iot": "arduino:samd:nano_33_iot",
+    "mkrvidor4000": "arduino:samd:mkrvidor4000",
+    "uno2018": "arduino:megaavr:uno2018",
+    "mkrnb1500": "arduino:samd:mkrnb1500",
+    "nanorp2040connect": "arduino:mbed_nano:nanorp2040connect",
+}
+
+
+# Runs arduino-cli, doesn't handle errors at all because am lazy
+def arduino_cli(cli_path, args=[]):
+    res = subprocess.run([cli_path, *args], capture_output=True, text=True)
+    return res.stdout
+
+
+# Generates file SHA256
+def sha2(file_path):
+    with open(file_path, "rb") as f:
+        return hashlib.sha256(f.read()).hexdigest()
+
+
+# Generate and copy loader Sketch binary data for specified board
+def create_loader_data(simple_fqbn, binary):
+    loader_path = f"firmwares/loader/{simple_fqbn}/loader.bin"
+    loader = Path(__file__).parent / loader_path
+    loader.parent.mkdir(parents=True, exist_ok=True)
+    shutil.copyfile(binary, loader)
+
+    file_hash = sha2(loader)
+
+    return {
+        "url": f"{DOWNLOAD_URL}/{loader_path}",
+        "checksum": f"SHA-256:{file_hash}",
+        "size": f"{loader.stat().st_size}",
+    }
+
+
+# Generate and copy all firmware binary data for specified board
+def create_firmware_data(simple_fqbn, binary, module, version):
+    binary_name = binary.name
+    firmware_path = f"firmwares/{module}/{version}/{binary_name}"
+    firmware = Path(__file__).parent / firmware_path
+    firmware.parent.mkdir(parents=True, exist_ok=True)
+    shutil.copyfile(binary, firmware)
+
+    file_hash = sha2(firmware)
+
+    return {
+        "version": version,
+        "url": f"{DOWNLOAD_URL}/{firmware_path}",
+        "checksum": f"SHA-256:{file_hash}",
+        "size": f"{firmware.stat().st_size}",
+    }
+
+
+def create_upload_data(fqbn, installed_cores):
+    upload_data = {}
+    # Assume we're on Linux
+    arduino15 = Path.home() / ".arduino15"
+
+    core_id = ":".join(fqbn.split(":")[:2])
+
+    # Get the core install dir
+    core = installed_cores[core_id]
+    (maintainer, arch) = core_id.split(":")
+    core_install_dir = (
+        arduino15 / "packages" / maintainer / "hardware" / arch / core["installed"]
+    )
+
+    with open(core_install_dir / "boards.txt") as f:
+        boards_txt = f.readlines()
+
+    board_upload_data = {}
+    for l in boards_txt:
+        board_id = fqbn.split(":")[2]
+        if l.startswith(f"{board_id}.upload"):
+            (k, v) = l.strip().split("=", maxsplit=1)
+            k = ".".join(k.split(".", maxsplit=1)[1:])
+            board_upload_data[k] = v
```

Comment:
Maybe this and code L98-102 could be put in a function since is basically the same code?